README.md
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/maria-yankova/fracture-fea-laf/HEAD)

# fracture-fea-laf

Mesh generation tool for standard fracture specimen models with a finite crack tip radius

Postprocessing functions for FE data that use a Weibull stress formulation for cleavage fracture prediction



CHANGELOG.md
# Change Log

## [0.1.1] - 2021.07.08

- Added mesh generation

## [0.1.0] - 2021.06.01

- Initial release


fracture_fea_laf/utils.py
import numpy as np

def polar2cart_2D(r, θ):
    """
    Convert 2D polar coordinates to Cartesian coordinates.

    """

    x = r * np.cos(θ)
    y = r * np.sin(θ)

    return x, y

def order_coplanar_points(points, normal, anticlockwise=True):
    """
    Find the clockwise or anticlockwise ordering of a set of coplanar 3D points.

    Parameters
    ----------
    points : ndarray of shape (3, N)
        The set of coplanar points (three-vector columns) whose ordering is to be found.
    normal : ndarray of shape (3, 1)
        Column three-vector representing the normal of the plane on which all points lie.

    Returns
    -------
    Ordered indices of points according to a clockwise or anticlockwise direction when
    looking in the opposite direction to `normal`.

    """

    # Normalise `normal` to a unit vector:
    normal = normal / np.linalg.norm(normal)

    # Compute the centroid:
    centroid = np.mean(points, axis=1)

    # Get the direction vectors from each point to the centroid
    p = points - centroid[:, np.newaxis]

    # Use the first point as the reference point
    # Find cross product of each point with the reference point
    crs = np.cross(p[:, 0:1], p, axis=0)

    # Find the scalar triple product of the point pairs and the normal vector
    stp = np.einsum('ij,ik->k', normal, crs)

    # Find the dot product of the point pairs
    dot = np.einsum('ij,ik->k', p[:, 0:1], p)

    # Find signed angles from reference point to each other point
    ang = np.arctan2(stp, dot)
    ang_order = np.argsort(ang)

    if not anticlockwise:
        ang_order = ang_order[::-1]

    return ang_order

def circle_points(r, n, centre):
    t = np.linspace(0, 2*np.pi, n)
    x = r * np.cos(t) + centre[0]
    y = r * np.sin(t) + centre[1]

    return np.c_[x, y]

def search_keep_order(A, B):
    sort_idx = A.argsort()
    out = sort_idx[np.searchsorted(A, B, sorter = sort_idx)]
    idx_layer_0 = np.nonzero(B[:,None] == A)[1]
    
    return idx_layer_0

fracture_fea_laf/postprocess.py
import numpy as np

def celsius2kelvin(temperature):
    return temperature + 273.15

def k2j(k, E, nu, plane_stress=False):
    """
    Convert fracture 
    
    Parameters
    ----------
    k: float
    E: float
        Young's modulus in GPa.
    nu: float
        Poisson's ratio
    plane_stress: bool
        True for plane stress (default) or False for plane strain condition.
    
    Returns
    -------
    float
    
    """
    
    if plane_stress:
        E = E / (1 - nu ** 2)
        
    return k ** 2 / E


def j2k(j, E, nu, plane_stress=True):
    """
    Convert fracture 
    
    Parameters
    ----------
    j: float (in N/mm)
    E: float
        Young's modulus in GPa.
    nu: float
        Poisson's ratio
    plane_stress: bool
        True for plane stress (default) or False for plane strain condition.
    
    Returns
    -------
    K : float
        Units are MPa m^0.5.
    """
    
    if plane_stress:
        E = E / (1 - nu ** 2)
        
    return (j * E) ** 0.5

def calc_cpf(data):
    data_ordered = np.sort(data)
    u =  (np.arange(len(data)) + 1 - 0.3) / (len(data) + 0.4)
    
    return data_ordered, u

def find_closest(a, b, x, tol=0.01):
    """
    For two arrays `a` and `b`, find the elements in `b` closest to the `x` element in `a`.
    
    Parameters
    ----------
    a: narray of shape (N,)
    b: narray of shape (N,)
    x: float
    tol: float
        Tolerance value. 
    Returns
    -------
    """
    a_idx = np.isclose(a, x, atol=tol)
    a_i = a[a_idx]
    b_i = b[a_idx]

    return a_i, b_i
def find_nearest(array, value):
    array = np.asarray(array)
    idx = (np.abs(array - value)).argmin()
    return idx

def Fk(shear_stress, temperature, Tath=700, Uk=0.33 * 1.60217657e-19, sigmap=900*1e6):
    """
    Parameters
    ----------
    shear_stress : ndarray of shape (N,)
        CRSS or maximum shear stress (in Pa or Nm-2)
    temperature : float
        Temperature (in Kelvin)
    Tath : float
        The athermal temperature in Kelvin. Default set for iron = 700 K.
    Uk : float
        The activation energy (in Joules). Default set for iron.
    sigmap : float
        critical Peierls stress (in Pascals). Default set for iron.

    Returns
    -------
    float
        Kink-formation free energy (in Joules).

    Ref. T. D. Swinburne, S. L. Dudarev, Kink-limited Orowan strengthening explains the brittle to ductile transition of irradiated and unirradiated bcc metals. Phys. Rev. Mater. 2, 73608 (2018).
    """

    fk = Uk *(1 - temperature / Tath - (shear_stress / sigmap) / (1 - temperature / Tath))
    fkneg_idx = np.where(fk<0)[0]

    fk[fkneg_idx] = 0.0
    
    return fk

def lstar(shear_stress, temperature, b=0.248e-9):
    """
    Parameters
    -------
    shear_stress : ndarray of shape (N,)
        CRSS or maximum shear stress (in Pa or Nm-2)
    temperature : float
        Temperature (in Kelvin)
    b : float
        Burger's vector (in m). Default set for iron. 
        
    Returns
    -------
    ndarray of shape (N,)
        Critical lenght L* (in m).
    
    """
    # Boltzmann constant
    kb =  1.3806498e-23 # J/K

    return b * np.exp( Fk(shear_stress, temperature) / (kb * temperature) )


def thinning_function(obstacle_distance, shear_stress, temperature):
    """
    Thinning function applied in a modified local approach to cleavage fracture predictions.
    
    Parameters
    ----------
    obstacle_distance: array_like
        Average distance between dislocation obstacles [1].
    shear_stress: array_like
        Maximum shear stress (in Pa or Nm-2)
    temperature: float 
        Temperature in degree Kelvin.
    Returns
    -------
    thinf : 

    References
    ----------
    [1] T. D. Swinburne and S. L. Dudarev, “Kink-limited Orowan strengthening explains the brittle to ductile transition of irradiated and unirradiated bcc metals,” Phys. Rev. Mater., vol. 2, p. 73608, 2018.
    
    """
    # Boltzmann constant
    kb = 1.3806498e-23 # Joules/Kelvin
    

    
    idx_more = lstar(shear_stress, temperature) > np.ones(len(shear_stress)) * obstacle_distance
    idx_less = ~idx_more
    ref_temp = - 200 + 273.15
    norm_factor = 1 / (
        2 * Fk(np.array([0]), ref_temp) / (kb * ref_temp)
    )[0]

    fk = np.zeros(len(shear_stress))
    fk[idx_more] = 2 * Fk(shear_stress, temperature)[idx_more] / (kb * temperature)
    fk[idx_less] = Fk(shear_stress, temperature)[idx_less] / (kb * temperature)


    thinf = norm_factor * fk
        
    return thinf


def calc_sw_diff(sw_data, n0, ri, mi, gdata_all, skipped_lines_all):
    
    sw0 = sw_data[n0][mi][gdata_all[n0]['j0_idx'] - np.array(skipped_lines_all[n0])]
    swi = sw_data[ri][mi][gdata_all[ri]['j0_idx'] - np.array(skipped_lines_all[ri])]
    
    return (swi - sw0) * 100 / swi

def pred_j0(pred_i, sw_data, idxm, jskipped_lines_all, gdata_all, temps, tol=0.002):
    """
    Predict the characteristic fracture toughness J0 based on the Weibull stress

    Parameters:
    -----------
    pred_i : integer
        Index basis for prediction (zero-based)
    sw_data :  a list of length i of lists of length Ni
        Weibull stress data for i temperatures and Ni time increments for each temperature 
    idxm : integer
        Index of shape parameter m to be used
    jskipped_lines_all : list
        Lines to be skipped in the fracture toughness data for each temperature due to no Weibull data, e.g. when J is very small and there is limited plasticity.
    gdata_all :  list of dicts
       Fracture toughness data for each temperature. Here, the 'j0' data is used from the dictionary.
    temps : ndarray of size (i,)
        Temperatures (in degrees Celsius)
    """
    sw_pred = sw_data[pred_i][idxm][gdata_all[pred_i]['j0_idx']-jskipped_lines_all[pred_i]]
 
    # print('Weibull stress: ', (sw_pred))
    
    for ri, sw_r in enumerate(sw_data):
        sw = np.array(sw_r[idxm])
        print('Temperature: ', str(temps[ri]))
        if ri != pred_i:
            j0_pred = np.where(np.isclose(sw, sw_pred, tol))
            shift_idx = len(gdata_all[ri]['gdata']['jintc']) - len(sw)
            # print('shift_idx: ', shift_idx)
            jpred = gdata_all[ri]['gdata']['jintc'][j0_pred[0]+shift_idx]

            if len(jpred)>0:
                print('Predicted J0 = {:4.2f}'.format(jpred.values[0]))
        else:
            print('basis for prediction')
            print('J0 = {:4.2f}'.format(gdata_all[pred_i]['j0']))

        print('###')

fracture_fea_laf/specimen_geometry.py
def standard_specimen(spec_type, dimensions='2D', fraction='half', aw_ratio=0.5):
    """
    Returns the dimensions of standard fracture specimens.

    Parameters
    ----------
    spec_type : string
        Geometry of the specimen, one of 'ct-1t' | 'charpy-senb-0.5' | 'charpy-senb-0.2'. More geometries can be added.
    dimensions : string
        One of '2D' | '3D'
    fraction : string
        One of 'quarter' | 'half' | 'full'
    aw_ratio : float
        Fixed at 0.5 for CT and variable for SENB.
    
    Returns
    -------
    if 2D:
        a/w, W, A, C, D, E, F
    if 3D:h
        a/w, W, A, B, C, D, E, F
    """
    
    specimens_dims = {
    'ct-1t':{
            'a/w': 0.5,
            'W': 50,
            'A': 62.5,
            'B': 25,
            'C': 12.5,
            'D': 23.5,
            'E': 60,
            'F': 37.5
        },
    'charpy-senb-0.5':{
            'a/w': 0.5,
            'W': 10,
            'L': 55,
            'B': 10,
    },
    'charpy-senb-0.1':{
            'a/w': 0.1,
            'W': 10,
            'L': 55,
            'B': 10,
    },
      'charpy-senb-0.2':{
            'a/w': 0.2,
            'W': 10,
            'L': 55,
            'B': 10,
    },
    }
    specimen = specimens_dims[spec_type]


    if spec_type=='ct-1t':
        if fraction=='half':
            specimen['E'] *= 0.5
        elif fraction=='quarter':
            specimen['E'] *= 0.5
            specimen['B'] *= 0.5
    elif spec_type[:4]=='char':
        if fraction=='half':
            specimen['L'] *= 0.5
        elif fraction=='quarter':
            specimen['L'] *= 0.5
            specimen['B'] *= 0.5
    
    if dimensions=='2D':
        specimen.pop('B')
        
    return specimens_dims[spec_type]
    

fracture_fea_laf/plot.py
from plotly import graph_objs as go

def plot_nodes(plot_x, plot_y, plot_z=None, indices=None, xrange=None, yrange=None,
               figsize=[400, 400], dticks=None, mode='markers+text',
              markersize=5, showlegend=False, show_labels=True, threed=False):
    """
    Parameters
    ----------
    show_labels : bool, optional
        If True, show node labels.
        
    """
    labels = ['node labels', 'element labels']
    labels = len(plot_x) * ['h']
    data = []
    if threed:
        if type(plot_x) == list:
            i = 0
            for plx, ply, plz in zip(plot_x, plot_y, plot_z):
                if indices == None:
                    text = None
                else:
                    text = indices[i]
                trace = {
                    'type': 'scatter3d',
                    'x': plx,
                    'y': ply,
                    'z': plz,
                     'mode': mode,
                    'marker': {'size': markersize},
                     'name': labels[i],                
                }
                if show_labels:
                    trace.update({'text': text, 'type': 'scatter3d'})

                data.append(trace)

                i += 1
        else:    
            if indices is not None:
                indices = indices
                print('here')
                print('indices shape: ', indices.shape)
                print('plot_x shape: ', plot_x.shape)
                data.append({
                     'type': 'scatter3d',
                        'x': plot_x,
                        'y': plot_y,
                        'z': plot_z,
                         'mode': mode,
                        'marker': {'size': 5},
                        'text': indices,
                        'name': labels[0],
                    })
            else:
                data.append({
                     'type': 'scatter3d',
                        'x': plot_x,
                        'y': plot_y,
                         'z': plot_z,
                        'text': None,
                         'mode': 'markers',
                        'marker': {'size': 5},
                        'name': labels[0],
                    })
    else:
        if type(plot_x) == list:
            i = 0
            for plx, ply in zip(plot_x,plot_y):
                if indices == None:
                    text = None
                else:
                    text = indices[i]
                trace = {
                    'type': 'scattergl',
                    'x': plx,
                    'y': ply,
                     'mode': mode,
                    'marker': {'size': markersize},
                    'name': labels[i],           
                }
                if show_labels:
                    trace.update({'text': text, 'type': 'scatter'})

                data.append(trace)

                i += 1
        else:    
            if indices is not None:
                indices = indices
                data.append({
                        'x': plot_x,
                        'y': plot_y,
                         'mode': mode,
                        'marker': {'size': 5},
                        'text': indices,
                         'name': labels[0],
                    })
            else:
                data.append({
                        'x': plot_x,
                        'y': plot_y,
                        'text': None,
                         'mode': 'markers',
                        'marker': {'size': 5},
                         'name': labels[0],
                    })

    layout = {
        'width': figsize[0],
        'height': figsize[1],
        'xaxis': {
            'zeroline': False,
            'showline': True,
            'mirror': 'all',
            'title': r'x (mm)',
            'titlefont':{'size': 13},
        },
        'yaxis': {
            'zeroline': False,
            'showline': True,
            'mirror': 'all',
            'title': r'y (mm)',
            'titlefont':{'size': 13},
            'scaleanchor': 'x',
            'scaleratio': 1,
        },
        # 'annotations':[
        # {
        #     'text':  r'$\LARGE{x}$',
        #     'x': 0.5,
        #     'y': -0.1,
        #     'xref': 'paper',
        #     'yref': 'paper',
        #     'showarrow': False,
        #     'font': {'size': 11},
        # },
        #     {
        #     'text':  r'$\LARGE{y}$',
        #     'x': -0.1,
        #     'y': 0.5,
        #     'xref': 'paper',
        #     'yref': 'paper',
        #     'showarrow': False,
        #     'font': {'size': 11},
        #     'textangle': -90
        # },
        # ],
        'legend':{
            'x': 0.6,
            'y': 0.7,
        },
         'margin': {
                    'l': 100,
                    'r': 20,
                    'b': 80,
                    't': 40,
         },

    }
    if xrange:
        layout['xaxis']['range'] = xrange
    if yrange:
        layout['yaxis']['range'] = yrange
    if dticks:
        layout['xaxis']['dtick'] = dticks[0]
        layout['yaxis']['dtick'] = dticks[1]
    layout['showlegend'] = showlegend
    f = go.FigureWidget(data, layout)
    return f
